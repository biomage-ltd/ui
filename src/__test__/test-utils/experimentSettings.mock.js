import initialState, {
  initialPipelineState,
  initialWorkerState,
} from '../../redux/reducers/experimentSettings/initialState';

const testableProcessingConfig = {
  cellSizeDistribution: {
    enabled: true,
    auto: true,
    filterSettings: {
      minCellSize: 10800,
      binStep: 200,
    },
  },
  mitochondrialContent: {
    enabled: true,
    auto: true,
    filterSettings: {
      method: 'absolute_threshold',
      methodSettings: {
        absolute_threshold: {
          maxFraction: 0.1,
          binStep: 200,
        },
      },
    },
  },
  classifier: {
    enabled: true,
    auto: true,
    filterSettings: {
      FDR: 0.1,
    },
  },
  numGenesVsNumUmis: {
    enabled: true,
    auto: true,
    filterSettings: {
      regressionType: 'gam',
      regressionTypeSettings: {
        gam: {
          'p.level': 0.00009,
        },
      },
    },
  },
  doubletScores: {
    enabled: true,
    auto: true,
    filterSettings: {
      probabilityThreshold: 0.2,
      binStep: 0.05,
    },
  },
  dataIntegration: {
    enabled: true,
    dataIntegration: {
      method: 'seuratv4',
      methodSettings: {
        seuratv4: {
          numGenes: 2000,
          normalisation: 'logNormalize',
        },
        fastmnn: {
          numGenes: 2000,
          normalisation: 'logNormalize',
        },
        unisample: {
          numGenes: 2000,
          normalisation: 'logNormalize',
        },
      },
    },
    dimensionalityReduction: {
      method: 'rpca',
      numPCs: 30,
      excludeGeneCategories: [],
    },
  },
  configureEmbedding: {
    enabled: true,
    embeddingSettings: {
      method: 'umap',
      methodSettings: {
        umap: {
          minimumDistance: 0.1,
          distanceMetric: 'euclidean',
        },
        tsne: {
          perplexity: 30,
          learningRate: 200,
        },
      },
    },
    clusteringSettings: {
      method: 'louvain',
      methodSettings: {
        louvain: {
          resolution: 0.5,
        },
      },
    },
  },
};

export default {
  ...initialState,
  processing: {
    ...initialState.processing,
    ...testableProcessingConfig,
  },
};

export { initialPipelineState, initialWorkerState };
