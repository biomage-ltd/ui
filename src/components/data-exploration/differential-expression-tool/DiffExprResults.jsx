import React, { useState, useEffect } from 'react';
import {
  useSelector,
  useDispatch,
} from 'react-redux';
import {
  Space, Button, Alert, Tooltip, Typography,
} from 'antd';
import Link from 'next/link';
import { LeftOutlined } from '@ant-design/icons';
import PropTypes from 'prop-types';
import _ from 'lodash';
import GeneTable from '../generic-gene-table/GeneTable';
import { geneTableUpdateReason } from '../../../utils/geneTable/geneTableUpdateReason';
import loadDifferentialExpression from '../../../redux/actions/differentialExpression/loadDifferentialExpression';

const { Text } = Typography;

const DiffExprResults = (props) => {
  const {
    experimentId, onGoBack, width, height,
  } = props;

  const dispatch = useDispatch();
  const loading = useSelector((state) => state.differentialExpression.properties.loading);
  const data = useSelector((state) => state.differentialExpression.properties.data);
  const total = useSelector((state) => state.differentialExpression.properties.total);
  const error = useSelector((state) => state.differentialExpression.properties.error);
  const comparisonGroup = useSelector((state) => state.differentialExpression.comparison.group);
  const comparisonType = useSelector((state) => state.differentialExpression.comparison.type);

  const properties = useSelector((state) => state.cellSets.properties);

  const [dataShown, setDataShown] = useState(data);
  const [exportAlert, setExportAlert] = useState(false);
  const [settingsListed, setSettingsListed] = useState(false);

  const columns = [
    {
      title: 'Avg log2FC',
      key: 'avg_log2FC',
      sorter: true,
    },
    {
      title: 'adj p-value',
      key: 'p_val_adj',
      sorter: true,
      render: (score, record) => <Tooltip title={`adj p-value: ${record.p_val_adj}`}>{score}</Tooltip>,
    },
    {
      title: 'Pct 1',
      key: 'pct_1',
      sorter: true,
      showSorterTooltip: {
        title: 'The percentage of cells where the feature is detected in the first group',
      },
    },
    {
      title: 'Pct 2',
      key: 'pct_2',
      sorter: true,
      showSorterTooltip: {
        title: 'The percentage of cells where the feature is detected in the second group',
      },
    },
  ];

  // When data changes, update rows.
  useEffect(() => {
    if (data && properties) {
      setDataShown(data);
    }
  }, [data, properties]);

  const isTableLoading = () => data.length === 0 || loading;

  const onUpdate = (newState, reason) => {
    // We handle `loading` and `loaded` in the HOC, no need to react to these.
    if (reason === geneTableUpdateReason.loaded || reason === geneTableUpdateReason.loading) {
      return;
    }

    dispatch(
      loadDifferentialExpression(
        experimentId,
        comparisonGroup[comparisonType],
        comparisonType,
        newState,
      ),
    );
  };

  const optionName = (word) => {
    const [rootGroup, clusterName] = word.split('/');

    let printString = '';

    if (!clusterName) {
      printString = _.capitalize(rootGroup);
    } else if (clusterName === 'rest') {
      printString = `Rest of ${properties[rootGroup].name}`;
    } else {
      printString = properties[clusterName]?.name || _.capitalize(clusterName);
    }

    return <Text strong>{printString}</Text>;
  };

  const { basis, cellSet, compareWith } = comparisonGroup[comparisonType];

  const renderExportAlert = () => {
    if (!exportAlert) return null;
    return (
      <Alert
        message={(
          <span>
            Exporting to CSV is not currently available here. Use the&nbsp;
            <Link
              target='_blank'
              as={`/experiments/${experimentId}/plots-and-tables/volcano`}
              href='/experiments/[experimentId]/plots-and-tables/volcano'
              passHref
            >
              {/* eslint-disable-next-line jsx-a11y/anchor-is-valid */}
              <a target='_blank'>volcano plot</a>
            </Link>
            &nbsp;in Plots and Tables to export results (opens in new tab).
          </span>
        )}
        type='info'
        closable
        showIcon
        afterClose={() => {
          setExportAlert(false);
        }}
      />
    );
  };

  return (
    <Space direction='vertical' style={{ width: '100%' }}>

      {/* This is needed so changes to the export alert don't cause the table to re-render. */}
      <Space direction='vertical' style={{ width: '100%' }}>
        <Button size='small' onClick={onGoBack}>
          <span>
            <LeftOutlined />
            Go back
          </span>
        </Button>
        {renderExportAlert()}
        <Button id='settingsButton' onClick={() => setSettingsListed(!settingsListed)}>
          {settingsListed ? 'Hide' : 'Show'}
          {' '}
          settings
        </Button>
      </Space>
      {settingsListed
        ? (
          <div id='settingsText'>
            {optionName(cellSet)}
            {' '}
            vs.
            {' '}
            {optionName(compareWith)}
            {' '}
            in
            {' '}
            {optionName(basis)}
          </div>
        ) : <div />}
      <GeneTable
        experimentId={experimentId}
        initialTableState={{
          sorter: {
            field: 'abszscore',
            columnKey: 'abszscore',
            order: 'descend',
          },
        }}
        onUpdate={onUpdate}
        columns={columns}
        loading={isTableLoading()}
        onExportCSV={() => { setExportAlert(true); }}
        error={error}
        width={width}
        height={height - 70 - (exportAlert ? 70 : 0) - (settingsListed ? 70 : 0)}
        data={dataShown}
        total={total}
      />
    </Space>
  );
};

DiffExprResults.defaultProps = {};

DiffExprResults.propTypes = {
  experimentId: PropTypes.string.isRequired,
  onGoBack: PropTypes.func.isRequired,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
};

export default DiffExprResults;
