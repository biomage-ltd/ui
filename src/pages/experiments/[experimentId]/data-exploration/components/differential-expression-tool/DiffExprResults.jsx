import React, { useState, useEffect } from 'react';
import {
  useSelector,
  useDispatch,
} from 'react-redux';
import {
  Space, Button, Alert, Tooltip,
} from 'antd';
import Link from 'next/link';
import { LeftOutlined } from '@ant-design/icons';
import PropTypes from 'prop-types';
import GeneTable from '../generic-gene-table/GeneTable';
import { geneTableUpdateReason } from '../../../../../../utils/geneTable/geneTableUpdateReason';
import loadDifferentialExpression from '../../../../../../redux/actions/differentialExpression/loadDifferentialExpression';

const DiffExprResults = (props) => {
  const {
    experimentId, onGoBack, cellSets, width, height,
  } = props;

  const dispatch = useDispatch();
  const loading = useSelector((state) => state.differentialExpression.properties.loading);
  const data = useSelector((state) => state.differentialExpression.properties.data);
  const total = useSelector((state) => state.differentialExpression.properties.total);
  const error = useSelector((state) => state.differentialExpression.properties.error);

  const [dataShown, setDataShown] = useState(data);
  const [exportAlert, setExportAlert] = useState(false);

  const columns = [
    {
      title: 'Z-score',
      key: 'zscore',
      sorter: true,
    },
    {
      title: 'Absolute Z-score',
      key: 'abszscore',
      sorter: true,
      render: (score, record) => <Tooltip title={`q-value: ${record.qval}`}>{score}</Tooltip>,
    },
    {
      title: 'log2 FC',
      key: 'log2fc',
      render: (num) => parseFloat(num.toFixed(1)),
      sorter: true,
    },
  ];

  // When data changes, update rows.
  useEffect(() => {
    if (data) {
      setDataShown(data);
    }
  }, [data]);

  const isTableLoading = () => data.length === 0 || loading;

  const onUpdate = (newState, reason) => {
    // We handle `loading` and `loaded` in the HOC, no need to react to these.
    if (reason === geneTableUpdateReason.loaded || reason === geneTableUpdateReason.loading) {
      return;
    }

    dispatch(
      loadDifferentialExpression(experimentId, cellSets, newState),
    );
  };

  const renderExportAlert = () => {
    if (!exportAlert) return null;
    return (
      <Alert
        message={(
          <span>
            Exporting to CSV is not currently available here. Use the&nbsp;
            <Link
              target='_blank'
              as={`/experiments/${experimentId}/plots-and-tables/volcano`}
              href='/experiments/[experimentId]/plots-and-tables/volcano'
              passHref
            >
              {/* eslint-disable-next-line jsx-a11y/anchor-is-valid */}
              <a target='_blank'>volcano plot</a>
            </Link>
            &nbsp;in Plots and Tables to dump results (opens in new tab).
          </span>
        )}
        type='info'
        closable
        showIcon
        afterClose={() => {
          setExportAlert(false);
        }}
      />
    );
  };

  return (
    <Space direction='vertical' style={{ width: '100%' }}>

      {/* This is needed so changes to the export alert don't cause the table to re-render. */}
      <Space direction='vertical' style={{ width: '100%' }}>
        <Button size='small' onClick={onGoBack}>
          <span>
            <LeftOutlined />
            Go back
          </span>
        </Button>
        {renderExportAlert()}
      </Space>

      <GeneTable
        experimentId={experimentId}
        initialTableState={{
          sorter: {
            field: 'abszscore',
            columnKey: 'abszscore',
            order: 'descend',
          },
        }}
        onUpdate={onUpdate}
        columns={columns}
        loading={isTableLoading()}
        onExportCSV={() => { setExportAlert(true); }}
        error={error}
        width={width}
        height={height - 70 - (exportAlert ? 70 : 0)}
        data={dataShown}
        total={total}
      />
    </Space>
  );
};

DiffExprResults.defaultProps = {};

DiffExprResults.propTypes = {
  experimentId: PropTypes.string.isRequired,
  onGoBack: PropTypes.func.isRequired,
  cellSets: PropTypes.object.isRequired,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
};

export default DiffExprResults;
