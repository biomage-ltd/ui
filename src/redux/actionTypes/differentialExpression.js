const DIFF_EXPR = 'diffExpr';

/**
 * Turns on the loading condition for differential expression.
 * Components that have the particular properties requested in their dependencies
 * should set a loading state at this event.
 */
const DIFF_EXPR_LOADING = `${DIFF_EXPR}/loading`;

/**
 * Sets the state of the differential expression data to be successfully loaded.
 * This is meant to be used for paginated data, where the data is
 * lazy-loaded by an API call or equivalent.
 */
const DIFF_EXPR_LOADED = `${DIFF_EXPR}/loaded`;

const DIFF_EXPR_ERROR = `${DIFF_EXPR}/error`;

/**
 * Sets the state for diffential expression settings
 *
 */
const DIFF_EXPR_COMPARISON_TYPE_SET = `${DIFF_EXPR}/comparisonType`;
const DIFF_EXPR_COMPARISON_GROUP_SET = `${DIFF_EXPR}/comparisonGroup`;

export {
  DIFF_EXPR_LOADING, DIFF_EXPR_LOADED, DIFF_EXPR_ERROR,
  DIFF_EXPR_COMPARISON_TYPE_SET, DIFF_EXPR_COMPARISON_GROUP_SET,
};
