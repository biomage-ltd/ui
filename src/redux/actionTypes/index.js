const GENE_LIST = 'GENE_LIST';
const LOAD_GENE_LIST = `${GENE_LIST}.LOAD`;
const UPDATE_GENE_LIST = `${GENE_LIST}.UPDATE`;

const SELECTED_GENES = 'SELECTED_GENES';

const BUILD_HEATMAP_SPEC = 'BUILD_HEATMAP_SPEC';
const UPDATE_GENE_EXPRESSION = 'UPDATE_GENE_EXPRESSION';
const UPDATE_HEATMAP_SPEC = 'UPDATE_HEATMAP_SPEC';

const SET_FOCUSED_GENE = 'SET_FOCUSED_GENE';

export {
  LOAD_GENE_LIST, UPDATE_GENE_LIST, SELECTED_GENES, BUILD_HEATMAP_SPEC,
  UPDATE_GENE_EXPRESSION, UPDATE_HEATMAP_SPEC,
  SET_FOCUSED_GENE,
};
