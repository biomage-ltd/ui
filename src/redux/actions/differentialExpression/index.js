import loadDifferentialGeneExpression from './loadDifferentialExpression';
import setComparisonType from './setComparisonType';
import setComparisonGroup from './setComparisonGroup';

export {
  loadDifferentialGeneExpression,
  setComparisonType,
  setComparisonGroup,
};
