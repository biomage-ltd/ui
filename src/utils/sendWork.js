import { v4 as uuidv4 } from 'uuid';
import moment from 'moment';
import connectionPromise from './socketConnection';
import WorkResponseError from './WorkResponseError';
import WorkTimeoutError from './WorkTimeoutError';
import getApiEndpoint from './apiEndpoint';

const sendWork = async (experimentId, timeout, body, requestProps = {}) => {
  const requestUuid = uuidv4();
  const io = await connectionPromise();

  // Check if we need to have a bigger timeout because the worker being down.
  const statusResponse = await fetch(`${getApiEndpoint()}/v1/experiments/${experimentId}/pipelines`);
  const jsonResponse = await statusResponse.json();

  const adjustedTimeout = (started && ready) ? timeout : timeout + 60;

  const { worker: { started, ready } } = jsonResponse;
  const timeoutDate = moment().add(adjustedTimeout, 's').toISOString();

  const request = {
    uuid: requestUuid,
    socketId: io.id,
    experimentId,
    timeout: timeoutDate,
    body,
    ...requestProps,
  };

  io.emit('WorkRequest', request);

  const responsePromise = new Promise((resolve, reject) => {
    io.on(`WorkResponse-${requestUuid}`, (res) => {
      const { response: { error } } = res;

      if (error) {
        return reject(new WorkResponseError(error, request));
      }

      return resolve(res);
    });
  });

  const timeoutPromise = new Promise((resolve, reject) => {
    const id = setTimeout(() => {
      clearTimeout(id);
      reject(new WorkTimeoutError(timeoutDate, request));
    }, adjustedTimeout * 1000);
  });

  return Promise.race([responsePromise, timeoutPromise]);
};

export default sendWork;
