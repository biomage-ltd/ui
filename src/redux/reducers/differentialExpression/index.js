import initialState from './initialState';
import {
  DIFF_EXPR_LOADING,
  DIFF_EXPR_LOADED,
  DIFF_EXPR_ERROR,
  DIFF_EXPR_COMPARISON_TYPE_SET,
  DIFF_EXPR_COMPARISON_GROUP_SET,
} from '../../actionTypes/differentialExpression';
import { EXPERIMENT_SETTINGS_PIPELINE_START } from '../../actionTypes/experimentSettings';

import differentialExpressionLoading from './differentialExpressionLoading';
import differentialExpressionLoaded from './differentialExpressionLoaded';
import differentialExpressionError from './differentialExpressionError';
import differentialExpressionSetType from './differentialExpressionSetType';
import differentialExpressionSetGroup from './differentialExpressionSetGroup';

const differentialExpressionReducer = (state = initialState, action) => {
  switch (action.type) {
    case DIFF_EXPR_LOADING: {
      return differentialExpressionLoading(state, action);
    }
    case DIFF_EXPR_LOADED: {
      return differentialExpressionLoaded(state, action);
    }
    case DIFF_EXPR_ERROR: {
      return differentialExpressionError(state, action);
    }
    case DIFF_EXPR_COMPARISON_TYPE_SET: {
      return differentialExpressionSetType(state, action);
    }
    case DIFF_EXPR_COMPARISON_GROUP_SET: {
      return differentialExpressionSetGroup(state, action);
    }
    case EXPERIMENT_SETTINGS_PIPELINE_START: {
      return initialState;
    }
    default: {
      return state;
    }
  }
};

export default differentialExpressionReducer;
