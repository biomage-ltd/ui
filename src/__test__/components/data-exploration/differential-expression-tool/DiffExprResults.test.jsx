import React from 'react';
import { mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import preloadAll from 'jest-next-dynamic';
import { act } from 'react-dom/test-utils';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import waitForActions from 'redux-mock-store-await-actions';

import DiffExprResults from '../../../../components/data-exploration/differential-expression-tool/DiffExprResults';
import sendWork from '../../../../utils/sendWork';
import { DIFF_EXPR_LOADING, DIFF_EXPR_LOADED } from '../../../../redux/actionTypes/differentialExpression';

jest.mock('localforage');
jest.mock('../../../../utils/sendWork', () => ({
  __esModule: true, // this property makes it work
  default: jest.fn(() => new Promise((resolve) => resolve({
    results: [
      {
        body: JSON.stringify({
          rows: [
            {
              p_val: 1.4969461240347763e-12, p_val_adj: 1.647289002209057e-11, avg_log2FC: -1.4274754343649423, gene_names: 'A',
            },
            {
              p_val: 2.4969461240347763e-12, p_val_adj: 2.647289002209057e-11, avg_log2FC: -2.4274754343649423, gene_names: 'B',
            },
            {
              p_val: 3.4969461240347763e-12, p_val_adj: 3.647289002209057e-11, avg_log2FC: -3.4274754343649423, gene_names: 'C',
            },
            {
              p_val: 4.4969461240347763e-12, p_val_adj: 4.647289002209057e-11, avg_log2FC: -4.4274754343649423, gene_names: 'D',
            },
            {
              p_val: 5.4969461240347763e-12, p_val_adj: 5.647289002209057e-11, avg_log2FC: -5.4274754343649423, gene_names: 'E',
            },
          ],
          total: 500,
        }),
      },
    ],
  }))),
}));

const mockStore = configureMockStore([thunk]);

const store = mockStore({
  genes: {
    selected: [],
  },
  cellInfo: {
    focus: {
      store: 'genes',
      key: 'C',
    },
  },
  differentialExpression: {
    properties: {
      data: [
        {
          p_val: 1.4969461240347763e-12, p_val_adj: 1.647289002209057e-11, avg_log2FC: -1.4274754343649423, gene_names: 'A',
        },
        {
          p_val: 2.4969461240347763e-12, p_val_adj: 2.647289002209057e-11, avg_log2FC: -2.4274754343649423, gene_names: 'B',
        },
        {
          p_val: 3.4969461240347763e-12, p_val_adj: 3.647289002209057e-11, avg_log2FC: -3.4274754343649423, gene_names: 'C',
        },
        {
          p_val: 4.4969461240347763e-12, p_val_adj: 4.647289002209057e-11, avg_log2FC: -4.4274754343649423, gene_names: 'D',
        },
        {
          p_val: 5.4969461240347763e-12, p_val_adj: 5.647289002209057e-11, avg_log2FC: -5.4274754343649423, gene_names: 'E',
        },
      ],
      loading: false,
      error: false,
      total: 5,
    },
    comparison: {
      type: 'between',
      group: {
        between: {
          cellSet: 'louvain/louvain-0',
          compareWith: 'louvain/louvain-1',
          basis: 'condition/condition-control',
        },
      },
    },
  },
});

describe('DiffExprResults', () => {
  beforeAll(async () => {
    await preloadAll();
  });

  beforeEach(() => {
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // deprecated
        removeListener: jest.fn(), // deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  });

  configure({ adapter: new Adapter() });
  it('renders correctly', () => {
    const component = mount(
      <Provider store={store}>
        <DiffExprResults experimentId='1234' onGoBack={jest.fn()} width={100} height={200} />
      </Provider>,
    );

    const table = component.find('Table Table');
    const spin = component.find('Table Spin');
    expect(spin.length).toEqual(1);
    expect(table.length).toEqual(1);
    expect(table.getElement().props.columns.length).toEqual(6);
    expect(table.getElement().props.columns[0].key).toEqual('lookup');
    expect(table.getElement().props.columns[1].key).toEqual('gene_names');
    expect(table.getElement().props.columns[2].key).toEqual('avg_log2FC');
    expect(table.getElement().props.columns[3].key).toEqual('p_val_adj');
    expect(table.getElement().props.columns[4].key).toEqual('pct_1');
    expect(table.getElement().props.columns[5].key).toEqual('pct_2');

    expect(table.getElement().props.dataSource.length).toEqual(5);
    expect(table.getElement().props.data.length).toEqual(5);
  });

  it('can sort the gene names in alphabetical order', async () => {
    const newPagination = {
      current: 1,
      pageSize: 4,
      showSizeChanger: true,
      total: 4,
    };

    const newSorter = {
      column: {
        dataIndex: 'gene_names',
        key: 'gene_names',
      },
      render: jest.fn(),
      columnKey: 'gene_names',
      field: 'gene_names',
      order: 'ascend',
    };

    const component = mount(
      <Provider store={store}>
        <DiffExprResults experimentId='1234' onGoBack={jest.fn()} width={100} height={200} />
      </Provider>,
    );

    const table = component.find('Table Table');

    act(() => {
      table.getElement().props.onChange(newPagination, {}, newSorter);
    });

    // // Wait for side-effect to propagate (properties loading and loaded).
    await waitForActions(store, [DIFF_EXPR_LOADING, DIFF_EXPR_LOADED]);

    expect(sendWork).toHaveBeenCalledWith('1234', 60,
      {
        cellSet: 'louvain-0',
        compareWith: 'louvain-1',
        basis: 'condition-control',
        experimentId: '1234',
        name: 'DifferentialExpression',
      },
      {
        pagination: {
          limit: 4, offset: 0, orderBy: 'gene_names', orderDirection: 'ASC', responseKey: 0,
        },
      });

    expect(store.getActions()[0]).toMatchSnapshot();
    expect(store.getActions()[1]).toMatchSnapshot();
  });

  it('Having a focused gene triggers focused view for `eye` button.', () => {
    // Redefine store from `beforeEach`.
    const component = mount(
      <Provider store={store}>
        <DiffExprResults experimentId='1234' onGoBack={jest.fn()} width={100} height={200} />
      </Provider>,
    );

    const table = component.find('Space Table Table');

    table.getElement().props.data.forEach((row) => {
      const lookupComponent = mount(
        <Provider store={store}>
          {row.lookup}
        </Provider>,
      );

      const focusButtonTooltip = lookupComponent.find('FocusButton Tooltip');

      if (row.gene_names === 'C') {
        expect(focusButtonTooltip.props().title).toContain('Hide');
      } else {
        expect(focusButtonTooltip.props().title).toContain('Show');
      }

      lookupComponent.unmount();
    });
  });
  it('Show comparison settings button works.', () => {
    const component = mount(
      <Provider store={store}>
        <DiffExprResults experimentId='1234' onGoBack={jest.fn()} width={100} height={200} />
      </Provider>,
    );
    const button = component.find('#settingsButton').first();
    expect(button.text()).toContain('Show');
    button.simulate('click');
    expect(button.text()).toContain('Hide');

    const div = component.find('#settingsText');
    expect(div.text()).toEqual('Louvain 0 vs. Louvain 1 in Condition control');
    button.simulate('click');
    expect(button.childAt(0).text()).toEqual('Show settings');
    expect(!div);
  });
});
