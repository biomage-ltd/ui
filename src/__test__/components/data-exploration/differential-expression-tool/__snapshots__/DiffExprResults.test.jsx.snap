// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiffExprResults can sort the gene names in alphabetical order 1`] = `
Object {
  "payload": Object {
    "experimentId": "1234",
  },
  "type": "diffExpr/loading",
}
`;

exports[`DiffExprResults can sort the gene names in alphabetical order 2`] = `
Object {
  "payload": Object {
    "cellSets": Object {
      "basis": "condition/condition-control",
      "cellSet": "louvain/louvain-0",
      "compareWith": "louvain/louvain-1",
    },
    "comparisonType": "between",
    "data": Array [
      Object {
        "avg_log2FC": -1.4274754343649423,
        "gene_names": "A",
        "p_val": 1.4969461240347763e-12,
        "p_val_adj": 1.647289002209057e-11,
      },
      Object {
        "avg_log2FC": -2.4274754343649425,
        "gene_names": "B",
        "p_val": 2.4969461240347763e-12,
        "p_val_adj": 2.647289002209057e-11,
      },
      Object {
        "avg_log2FC": -3.4274754343649425,
        "gene_names": "C",
        "p_val": 3.4969461240347765e-12,
        "p_val_adj": 3.647289002209057e-11,
      },
      Object {
        "avg_log2FC": -4.4274754343649425,
        "gene_names": "D",
        "p_val": 4.496946124034777e-12,
        "p_val_adj": 4.647289002209057e-11,
      },
      Object {
        "avg_log2FC": -5.4274754343649425,
        "gene_names": "E",
        "p_val": 5.4969461240347765e-12,
        "p_val_adj": 5.647289002209057e-11,
      },
    ],
    "experimentId": "1234",
    "total": 500,
  },
  "type": "diffExpr/loaded",
}
`;
